In [1]:
"""
These tests are an important sanity check

If any modification ruins the test functionality then we can be sure 
that the modification is detrimental
"""

import numpy as np
import matplotlib.pyplot as plt

# Add Backend to path
import sys

"""
Change Path according to own machine
"""

#sys.path.append('C:/Users/eeveetza/Documents/GitHub/AlgoBrain/src/com/company/Backend')
sys.path.append('C:/Users/alexa/OneDrive/Documents/GitHub/AlgoBrain/src/com/company/Backend')
#sys.path.append('C:/Users/Tbuob/OneDrive/Dokumente/GitHub/AlgoBrain/src/com/company/Backend')

from Environment import World
from Experiment import Experiment

"""
Mapping integers - instructions:
    
0  <==> NopA
1  <==> NopB
2  <==> NopC
3  <==> IfNEq
4  <==> IfLess
5  <==> Swap
6  <==> Pop
7  <==> Push
8  <==> SwapStack
9  <==> RightShift
10 <==> LeftShift
11 <==> Inc
12 <==> Dec
13 <==> Add
14 <==> Sub
15 <==> Nand
16 <==> HAlloc
17 <==> HDivide
18 <==> IO
19 <==> HCopy
20 <==> HSearch
21 <==> MovHead
22 <==> JmpHead
23 <==> GetHead£
24 <==> SetFlow
25 <==> IfLabel

"""

# clears all registers (under the assumption that the active stack is empty)
clear = [6, 0, 6, 1, 6, 2]

# Custom made test instructions
program_not = [6, 1, 18, 2, 14,0, 12, 0, 18, 0] + clear
program_nand = [18,1,18,2,15,0,18,0] + clear
program_and = [18,1,18,2,15,2,6,1,14,0,12,0,18,0] + clear
program_or = [6, 1, 18, 2, 14, 0, 12, 0, 6, 1, 18, 2, 14, 2, 12, 2, 7, 0, 6, 1, 15, 0, 18, 0] + clear
program_or_n = [6, 1, 18, 2, 14, 2, 12, 2, 18, 1, 15, 0, 18, 0] + clear
program_and_n = [6, 1, 18, 2, 14, 2, 12, 2, 18, 1, 15, 2, 6, 1, 14, 0, 12, 0, 18, 0] + clear
program_nor = [18, 2, 14, 0, 12, 0, 6, 1, 18, 2, 14, 2, 12, 2, 7, 0, 6, 1, 15, 2, 6, 1, 14, 0, 12, 0, 18, 0] + clear
program_xor = [18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15,\
           2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0] + clear
program_equ = [18, 1, 18, 2, 7, 1, 7, 2, 8, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7,\
               2, 7, 0, 6, 2, 14, 2, 12, 2, 6, 1, 15, 2, 8, 6, 1, 15, 0, 18, 0]

In [ ]:
#%%
"""
Parser to create instructions by reading input stream

All instructions to be written lowercase without dashes
"""

result = []
instr = []
while True:
    data = input("Instructions: ")
    if data == "end":
            print("result = " + str(result))
            break
    else:
        number_of_empty_responses = 0
        
        # Boring case checking
        if data == "nopa" or data == "a":
            result.append(0)
            instr.append("nopa")
        elif data == "nopb" or data == "b":
            result.append(1)
            instr.append("nopb")
        elif data == "nopc" or data == "c":
            result.append(2)
            instr.append("nopc")
        elif data == "ifneq":
            result.append(3)
        elif data == "ifless":
            result.append(4)
        elif data == "swap":
            result.append(5)
        elif data == "pop":
            result.append(6)
        elif data == "push":
            result.append(7)
        elif data == "swapstack":
            result.append(8)
        elif data == "rshift":
            result.append(9)
        elif data == "lshift":
            result.append(10)
        elif data == "inc":
            result.append(11)
        elif data == "dec":
            result.append(12)
        elif data == "add":
            result.append(13)
        elif data == "sub":
            result.append(14)
        elif data == "nand":
            result.append(15)
        elif data == "halloc":
            result.append(16)
        elif data == "hdivide":
            result.append(17)
        elif data == "io":
            result.append(18)
        elif data == "hcopy":
            result.append(19)
        elif data == "hsearch":
            result.append(20)
        elif data == "movhead":
            result.append(21)
        elif data == "jmphead":
            result.append(22)
        elif data == "gethead":
            result.append(23)
        elif data == "setflow":
            result.append(24)
        elif data == "iflabel":
            result.append(25)
        elif data == "consume":
            result.append(26)
        elif data == "back":
            result.pop(-1)
        elif data == "clear":
            result = []
        print(result)

In [2]:
#%%
"""
Test 1:
    
    An organism that computes NOT
    
    NOT(A) == -A - 1
    
    POP-B
    IO-C
    SUB-A
    DEC-A
    IO-A
    
"""
instructions = program_not

world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0,notify_ = True)

world.place_custom(program_not)

world.schedule(13)


EMULATOR AT POSITION (0, 0) COMPUTED NOT
INPUT:  3200118152
OUTPUT: 1094849143


In [3]:
#%%
"""
Test 2:
    
    An organim that computes NAND
    
    IO-B
    IO-C
    NAND-A
    IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_nand)

world.schedule(10)


EMULATOR AT POSITION (0, 0) COMPUTED NAND
INPUT: [3387003204  206414249]
OUTPUT:  4156456703


In [5]:
"""
Test 3:
    
    An organism that computes AND
    
    Serial connection NAND => NOT
    
    AND(A,B) = NOT(NAND(A,B))
    
    IO-B
    IO-C
    NAND-C
    POP-B
    SUB-A
    INC-A
    IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_and)

world.schedule(100)


EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [2323369136 1614133869]
OUTPUT:  3246112

EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [4042925046 3772234831]
OUTPUT:  3771867206

EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [2151040989 2844560488]
OUTPUT:  2147746888

EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [1855890127 3151919130]
OUTPUT:  715005962

EMULATOR AT POSITION (0, 0) COMPUTED AND
INPUT: [ 313977300 1413214760]
OUTPUT:  271769600


In [5]:
"""
Test 4:
    
    An organism that computes OR
    
    OR(A,B) == NOT(NOT(A) AND NOT(B))
    
    NOT(A) AND NOT(B) = NOT(NAND(NOT(A),NOT(B)))
    
    All together:
    
        OR(A,B) = NAND(NOT(A),NOT(B))
        
    
    #1: 
        Save NOT(A) in regA
        
        POP-B
        IO-C
        SUB-A
        DEC-A
        
    #2:
        Save NOT(B) in regC
        
        POP-B
        IO-C
        SUB-C
        DEC-C
        
    #3:
        Move regA to regB
        
        PUSH-A
        POP-B
    
    #4:
        NAND to regA, IO-A
        
        NAND-A
        IO-A
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_or)

world.schedule(15)


EMULATOR AT POSITION (0, 0) COMPUTED OR
INPUT: [2131142986 1842429885]
OUTPUT:  2144845823


In [5]:
"""
Test 5:
    
    An organism that computes OR_N
    
    Take A OR NOT(B)
    
    A OR NOT(B) = NOT(A) NAND B
    
    # Save NOT(A) to C
    # Input B
    # NAND to A
    # IO-A
    
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_or_n)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED OR_N
INPUT: [2545654717  212422572]
OUTPUT:  1827500014


In [7]:
"""
Test 6:
    
    An organism that computes AND_N
    
    Take A AND NOT(B)
    
    a and not(b) = not(a nand not(b))
    
    # Save NOT(B) to regC
    # Load A to regB
    # NAND-C
    # POP-B
    # SUB-A
    # DEC-A
    # IO-A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_and_n)

world.schedule(12)


EMULATOR AT POSITION (0, 0) COMPUTED AND_N
INPUT: [ 924983309 3209686614]
OUTPUT:  2097161


In [8]:
"""
Test 7:
    
    An organism that computes NOR
    
    NOT(A OR B)
    
    A OR B = NAND(NOT(A),NOT(B))
    
    How to do it:
        
    #1: 
        Save NOT(A) in regA
        
        IO-C
        SUB-A
        DEC-A
        
    #2:
        Save NOT(B) in regC
        
        POP-B
        IO-C
        SUB-C
        DEC-C
        
    #3:
        Move regA to regB
        
        PUSH-A
        POP-B
    
    #4:
        NAND to regC
        
        NAND-C
        
    #5: Clear B
        POP-B
        
    #6: Sub to A, Dec A, IO A
    
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_nor)

world.schedule(20)


EMULATOR AT POSITION (0, 0) COMPUTED NOR
INPUT: [3045857021   81714495]
OUTPUT:  1243684864


In [9]:
"""
Test 8:
    
    An organism that computes XOR
    
    (A and ~B) or (~A and B)
    
    NOT(NAND(A,NOT(B))) OR NOT(NAND(NOT(A),B))
    
    This ends up being equivalent to
    
    nand(nand(a,not(b)),nand(not(a),b))
    
    #1:
        input one to a
        input other to c
        push both to stack for later use. push 2nd input first and then the first one
        switch stack for popping
        calculate not(other) as:
            sub-c
            dec-c
        move first one from a to b by doing:
            push-a
            pop-b
            pop-a
        
        now we have:
            reg_a = 0
            reg_b = Input0
            reg_c = not(Input1)
            stack = (b,a)
        we just nand this to c
        pop-b
        push-c
        
        repeat first part
        
        swap stack
        pop a
        pop c
        
        Detailed Guide:
            
            # Inputing and saving to stack:
                IO-A
                IO-C
                PUSH-A
                PUSH-C
                
            Stack0 now is [i1,i0]
            
            Registers are:
                A = i0
                B = 0
                C = i1
                
            swapstack
            
            # Calculating not(i1):
                sub-c
                dec-c
                
            # Move one from A to B:
                push-a
                pop-b
                pop-a
                
                
            Registers are:
                
                regA = 0
                regB = A
                regC = NOT(B)
                
            # Nand them to C:
                
                nand-c
                
            # pop b to clear it:
                pop-b
            
            push-c
            
            now the second stack has nand(a,not(b)) in it.
            Switching to first stack with [b,a] in it
            
            swapstack
            pop-a
            pop-c
            
            # do not of register C which contains A
            
            sub-c
            dec-c
            
            # move B from A to regB
            
            push-a
            pop-b
            pop-a
            
            # Registers now:
                
                regA = 0
                regB = B
                regC = not(A)
            
            # nand them to c
            nand-c
            
            # swap to other stack
            
            swapstack
            
            # retrieve first result to regb
            
            pop-b
            
            # nand
            nand-a
            
            io-a
        
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_xor)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED XOR
INPUT: [2890357373 2208707842]
OUTPUT: 803303295


In [2]:
"""
Test EQU:
    
    An organism that computes EQU
    
    EQU(A,B) = (A and B) or (~A and ~B)
    
    not(nand(a,b)) or not(nand(not(a),not(b)))
    
    nand(nand(a,b),nand(not(a),not(b)))
    
    #0: Input a,b; store in active stack
    #1: Switch stack
    #2: Compute nand(a,b)
    #3: Clear register b
    #4: Store nand(a,b) to stack
    #5: Switch stack
    #6: Put a to reg_a, b to reg_c
    #7 : Calculate not(b) in reg_c, store to stack
    #8: Move a to reg_c by doing push-a pop-c
    #9: Calculate not(a) in reg_c
    #10: pop not(b) to reg_b
    #11: nand them to c
    #12: swap stack
    #13: pop to b
    #14: nand to a
    #15: io-a
        
"""
world = World(1, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_equ)

world.schedule(30)


EMULATOR AT POSITION (0, 0) COMPUTED EQU
INPUT: [ 350579222 1437327484]
OUTPUT: 3199301013


In [11]:
"""
Test 9:
    
    Move Up
    
"""

program = [27,0,0,1,2]

world = World(2, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program,(1,0))

print("Before Moving Up")
print(world.pool.get())


world.schedule(1)

print("\nAfter Moving Up")
print(world.pool.get())

Before Moving Up
[[0 0]
 [<DigitalOrganism.CPUEmulator object at 0x000001B1527028E0> 0]]

After Moving Up
[[<DigitalOrganism.CPUEmulator object at 0x000001B1527028E0> 0]
 [0 0]]


In [2]:
"""
Test 10:
    
    Move Down
    
"""

program = [28,0,0,1,2]

world = World(2, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program,(0,0))

print("Before Moving Down")
print(world.pool.get())


world.schedule(1)

print("\nAfter Moving Down")
print(world.pool.get())

Before Moving Down
[[<DigitalOrganism.CPUEmulator object at 0x00000187D9200FA0> 0]
 [0 0]]

After Moving Down
[[0 0]
 [<DigitalOrganism.CPUEmulator object at 0x00000187D9200FA0> 0]]


In [13]:
"""
Test 11:
    
    Move Left
    
"""

program = [29,0,0,1,2]

world = World(2, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program,(0,1))

print("Before Moving Left")
print(world.pool.get())


world.schedule(1)

print("\nAfter Moving Left")
print(world.pool.get())

Before Moving Left
[[0 <DigitalOrganism.CPUEmulator object at 0x000001B152708460>]
 [0 0]]

After Moving Left
[[<DigitalOrganism.CPUEmulator object at 0x000001B152708460> 0]
 [0 0]]


In [14]:
"""
Test 12:
    
    Move Right
    
"""

program = [30,0,0,1,2]

world = World(2, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0)

world.place_custom(program,(0,0))

print("Before Moving Right")
print(world.pool.get())


world.schedule(1)

print("\nAfter Moving Right")
print(world.pool.get())

Before Moving Right
[[<DigitalOrganism.CPUEmulator object at 0x000001B152702220> 0]
 [0 0]]

After Moving Right
[[0 <DigitalOrganism.CPUEmulator object at 0x000001B152702220>]
 [0 0]]


In [15]:
"""
Test 13:
    
    Consume
"""

print("A demonstration of the Consume instruction:\n")

program_predator = [26] + (len(program_xor))*[0]
program_prey = program_xor

world = World(2, replacement_strategy = "neighborhood", cm_prob = 0, ins_prob = 0, del_prob = 0, notify_ = True)

world.place_custom(program_predator, position = (0,0))
world.place_custom(program_prey, position = (0,1))

print(world)

world.schedule(1)

print("An emulator has now been consumed. The predator can now output XOR, a powerful function\n")
print(world)

print("Running for a further 30 cycles produces:")

world.schedule(30)

A demonstration of the Consume instruction:

Emulator 1: 

Register A: 0
Register B: 0
Register C: 0
Instruction Pointer: 0
Memory: [26, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Age: 0

Emulator 2: 

Register A: 0
Register B: 0
Register C: 0
Instruction Pointer: 0
Memory: [18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2]
Age: 0


An emulator has now been consumed. The predator can now output XOR, a powerful function

Emulator 1: 

Register A: 0
Register B: 0
Register C: 0
Instruction Pointer: 1
Memory: [26, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

# Debugging:

### 1: Find out what those organisms of length 10 are

In [16]:
world = World(2)
world.place_custom([10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1,\
            18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1])

In [33]:
def min_length(world):
    
    len_ = 100000
    
    for i in range(world.pool.shape[0]):
        for j in range(world.pool.shape[1]):
            if world.pool.get((i,j)) == 0:
                pass
            else:
                if len(world.pool.get((i,j)).original_memory) < len_:
                    len_ = len(world.pool.get((i,j)).original_memory)
                    
    return len_

In [34]:
def max_length(world):
    
    len_ = -1
    
    for i in range(world.pool.shape[0]):
        for j in range(world.pool.shape[1]):
            if world.pool.get((i,j)) == 0:
                pass
            else:
                if len(world.pool.get((i,j)).original_memory) > len_:
                    len_ = len(world.pool.get((i,j)).original_memory)
                    
    return len_

In [18]:
min_length(world)

50

In [ ]:
while min_length(world) > 10:
    world.schedule(50)

In [22]:
min_length(world)

60

In [23]:
print(world.pool.get((1,1)).original_memory)

[20, 1, 17, 0, 16, 21, 2, 1, 23, 11, 19, 17, 22, 23, 9, 20, 6, 7, 11, 9, 2, 25, 25, 24, 14, 24, 18, 11, 3, 11, 25, 3, 12, 18, 24, 18, 19, 15, 25, 1, 21, 17, 19, 0, 19, 3, 19, 25, 23, 19, 20, 14, 23, 19, 25, 0, 17, 19, 22, 0, 1]


Track its lineage:

In [13]:
import numpy as np

with open('divlog2.log') as f:
    content = f.readlines()
    
con = np.array(content)

In [14]:
current = '[10, 20, 2, 0, 21, 2, 5, 0, 22, 6, 16, 12, 19, 20, 9, 17, 10, 3, 5, 20, 19, 23, 10, 14, 5, 3, 14, 3, 12, 21, 2, 8, 2]\n'

lineage = []
lineage.append(current)

found_start = False

loc_old = -1
loc_new = -1

while not found_start:
    # Find the last occurence of the current organism as a child
    # If all of the occurences of the current organism are as parent, we have found the start of the lineage
    
    loc_old = loc_new
    
    temp = 0
    location = np.where(con == current)[0][-1]
    
    for i in np.where(con == current)[0]:
        if con[i+1] == "\n" and i <= location:
            location = i
            temp = 1
            
    if temp == 0:
        found_start = True
        
    loc_new = location
    
    if loc_old == loc_new:
        found_start = True
    else:
        # Find the parent, append it to lineage
        lineage.append(con[location-1])
    
        # current is now the parent
        current = con[location-1]

lineage.reverse()

print("Full NAND lineage:\n")

for element in lineage:
    print(element)

Full NAND lineage:

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 13, 3, 23, 17, 7, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 18, 15, 11, 16, 0, 13, 7, 9, 2, 18, 17, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 8, 7, 13, 3, 23, 17, 3, 7, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 2, 11, 11, 16, 0, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 13, 9, 13, 7, 12, 3, 23, 17, 3, 9, 15, 20, 17, 22, 9, 18, 2, 12, 14, 2, 1, 4, 11, 11, 16, 0, 3, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 15, 9, 13, 7, 12, 3, 23, 17, 9, 9, 15, 20, 17, 22, 3, 18, 2, 12, 14, 2, 1, 4, 11, 11, 16, 16, 0, 3, 13, 7, 17, 2, 18, 19, 20, 3, 19, 19, 25, 2, 0, 17, 21, 0, 1]

[10, 20, 2, 0, 21, 2, 15, 9, 3, 7, 12, 3, 23, 17, 9, 9, 15, 20, 17, 22, 3, 18, 2

Parsing mutations in the new lineage tracking system:



In [14]:
start = [16, 20, 2, 0, 21, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]
mutations = [['D', 2]]

In [15]:
for mutation in mutations:
    
    if mutation[0] == "C":
        start[mutation[1]] = mutation[2]
        
    if mutation[0] == "I":
        start.insert(mutation[1], mutation[2])
        
    if mutation[0] == "D":
        del start[mutation[1]]

# EXPERIMENTS:

### Default -> Not, run 1:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.0025, N=30
- Instruction Set: Default
- Runtime: 10min 58s
- Program: [16, 15, 22, 0, 21, 2, 2, 2, 2, 2, 19, 9, 2, 12, 2, 2, 4, 4, 2, 2, 2, 17, 2, 2, 2, 21, 1, 2, 2, 17, 6, 2, 18, 2, 17, 2, 4, 8, 2, 2, 2, 15, 18, 20, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]
- Program Length: 54
- Number of Mutations: 46, of which:
-   24 Copy
-   11 Insertion
-   11 Deletion

#### Interesting Remark: Even though there was an equal number of insertion and deletion mutations, likely due to a change in the flow of the program, the end organism ended up being longer than the original organism (length 50)

All Mutations:

[['D', 18],
 ['C', 12, 8],
 ['I', 41, 19],
 ['C', 25, 9],
 ['C', 2, 22],
 ['D', 28],
 ['C', 28, 10],
 ['I', 40, 19],
 ['C', 1, 19],
 ['C', 32, 20],
 ['I', 23, 22],
 ['C', 31, 18],
 ['I', 16, 4],
 ['C', 26, 18],
 ['D', 17],
 ['D', 34],
 ['C', 41, 11],
 ['D', 11],
 ['D', 24],
 ['I', 17, 4],
 ['C', 34, 5],
 ['D', 18],
 ['C', 38, 15],
 ['I', 36, 18],
 ['C', 34, 17],
 ['D', 44],
 ['D', 7],
 ['I', 36, 5],
 ['C', 35, 4],
 ['D', 32],
 ['C', 25, 1],
 ['I', 39, 18],
 ['C', 10, 19],
 ['C', 11, 9],
 ['I', 28, 17],
 ['C', 29, 6],
 ['I', 13, 12],
 ['C', 37, 8],
 ['I', 41, 15],
 ['C', 47, 2],
 ['D', 43],
 ['C', 21, 17],
 ['C', 40, 2],
 ['C', 25, 21],
 ['C', 5, 2],
 ['C', 1, 15]]
 
 Mutations to be read as: [TYPE, LOCATION, INSTRUCTION]
 
 For Deletion Mutations the "INSTRUCTION" element is omitted

### Default -> Not, run 2:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtimes: 56s
- Program Length: 51
- Number of Mutations: 64, of which:
-   61 Copy
-   2 Insertion
-   1 Deletion

#### NOTE: With a higher copy probability the runtime is reduced by a factor of 10. This is subject to randomness.

### Default -> Not, run 3:


- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtimes: 36s
- Program Length: 51
- Number of Mutations: 64, of which:
-   61 Copy
-   2 Insertion
-   1 Deletion

#### NOTE: With a higher copy probability the runtime is reduced by a factor of 10. This is subject to randomness.

### Default to NOT, run 4:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 1min 6s
- Program Length: 49
- Number of Mutations: 23, of which:
-   22 Copy
-   0 Insertion
-   1 Deletion


### NOT -> OR:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 15min 31s
- Program Length: 52
- Number of Mutations: 212, of which:
-   202 Copy
-   5 Insertion
-   5 Deletion

### Default -> NOR:

- Parameters: ins_prob = 0.05, del_prob = 0.05, cm_prob = 0.05, N=30
- Instruction Set: Default
- Runtime: 27min 11s
- Program Length: 44
- Number of Mutations:
- Copy: 738
- Insert: 15
- Delete: 21
- TOTAL: 774

Only computes NOR

###

23min for NOT -> NOR

0.05 ins
0.05 del
0.01 cm
custom IS

# RUN EXPERIMENT HERE:

In [6]:
%%time

# Define the size of the pool here:
N = 30

# Define the start, targ£et and parameters of the experiment

# stat_cycles determines how often you're informed of the statistics of the population

experiment = Experiment(start_organism="default", target_function="nor", ins_prob = 0.05,\
                        del_prob = 0.05, cm_prob = 0.05,N=N, stat_cycles = 500, instruction_set = "custom")

Wall time: 214 ms


In [7]:
%%time

# Runs the experiment and looks for the target function
experiment.run()

# Captures the emulator which computes the functions realiably for further analysis
em = None

counter = 0

for i in range(N):
    for j in range(N):
        if experiment.world.pool.get((i,j)).original_memory == experiment.first_specimen:
            counter += 1
            em = experiment.world.pool.get((i,j))



Min Length: (48, (5, 28))
Max Length: (52, (6, 16))
Mean Length: 50.05111111111111
Min Rate: 48
Max Rate: 52
Mean Rate: 50.05111111111111
Max Age: 500
Mean Age: 163.35444444444445
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (45, (6, 23))
Max Length: (52, (6, 15))
Mean Length: 49.77
Min Rate: 45
Max Rate: 52
Mean Rate: 49.77
Max Age: 759
Mean Age: 263.25444444444446
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (45, (6, 23))
Max Length: (52, (1, 11))
Mean Length: 49.544444444444444
Min Rate: 45
Max Rate: 384
Mean Rate: 50.63333333333333
Max Age: 1019
Mean Age: 286.08222222222224
NOT: 0
NAND: 0
AND: 0
OR_N: 0
OR: 0
AND_N: 1
NOR: 0
XOR: 0
EQU: 0


Min Length: (43, (10, 2))
Max Length: (53, (1, 10))
Mean Length: 49.492222222222225
Min Rate: 43
Max Rate: 384
Mean Rate: 52.72
Max Age: 4733
Mean Age: 318.49555555555554
NOT: 1
NAND: 0
AND: 1
OR_N: 0
OR: 0
AND_N: 1
NOR: 0
XOR: 0
EQU: 0


Min Length: (45, (2, 21))
Max Length



Min Length: (43, (24, 11))
Max Length: (55, (7, 23))
Mean Length: 49.33222222222222
Min Rate: 43
Max Rate: 106
Mean Rate: 49.44888888888889
Max Age: 1194
Mean Age: 300.51444444444445
NOT: 0
NAND: 1
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (41, (4, 21))
Max Length: (55, (10, 23))
Mean Length: 49.34222222222222
Min Rate: 41
Max Rate: 1536
Mean Rate: 58.35
Max Age: 5440
Mean Age: 317.2288888888889
NOT: 1
NAND: 0
AND: 1
OR_N: 0
OR: 0
AND_N: 1
NOR: 0
XOR: 0
EQU: 0
Reacting on nor


Min Length: (42, (3, 24))
Max Length: (55, (10, 19))
Mean Length: 49.364444444444445
Min Rate: 42
Max Rate: 24064
Mean Rate: 86.01666666666667
Max Age: 30080
Mean Age: 436.31666666666666
NOT: 1
NAND: 0
AND: 1
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


Min Length: (42, (1, 25))
Max Length: (55, (7, 23))
Mean Length: 49.37
Min Rate: 42
Max Rate: 24064
Mean Rate: 82.40333333333334
Max Age: 56576
Mean Age: 483.17555555555555
NOT: 1
NAND: 0
AND: 1
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 

KeyboardInterrupt: 

In [10]:
em = experiment.world.pool.get()[experiment.world.rates == 24064][0]

In [12]:
experiment.display_statistics()

Min Length: (35, (3, 22))
Max Length: (52, (8, 7))
Mean Length: 49.76555555555556
Min Rate: 35
Max Rate: 104
Mean Rate: 59.964444444444446
Max Age: 734
Mean Age: 157.68444444444444
NOT: 132
NAND: 1
AND: 0
OR_N: 0
OR: 0
AND_N: 0
NOR: 0
XOR: 0
EQU: 0


In [6]:
"""
A script to count the mutation types
"""

c = 0
ins = 0
d = 0

for i in range(len(em.mutations)):
    
    if em.mutations[i][0] == "C":
        c += 1
    elif em.mutations[i][0] == "I":
        ins += 1
    elif em.mutations[i][0] == "D":
        d += 1
        
print("Copy: " + str(c))
print("Insert: " + str(ins))
print("Delete: " + str(d))
print("TOTAL: " + str(c+d+ins))

Copy: 27
Insert: 11
Delete: 2
TOTAL: 40


In [15]:
"""
A script to test the evolved organism
"""

world = World(1,notify_ = True)
world.place_custom(em.original_memory)
world.schedule()


EMULATOR AT POSITION (0, 0) COMPUTED NOR
INPUT: [2006504471 2884074488]
OUTPUT:  0


KeyboardInterrupt: 

In [20]:
em.fun_and_n

True

In [10]:
"""
A Script to recreate the program from the ancestor and the mutations
"""

start = em.ancestor.copy()
mutations = em.mutations.copy()

for mutation in mutations:
    
    if mutation[0] == "C":
        start[mutation[1]] = mutation[2]
        
    if mutation[0] == "I":
        start.insert(mutation[1], mutation[2])
        
    if mutation[0] == "D":
        del start[mutation[1]]
        
print(start == experiment.first_specimen)

True


In [11]:
len(experiment.first_specimen)

79

In [12]:
"""
A list of all the mutations which led to the final organism
"""
mutations

[['C', 2, 23, 0],
 ['C', 44, 19, 0],
 ['C', 44, 9, 19],
 ['C', 60, 1, 19],
 ['C', 50, 3, 45],
 ['I', 63, 19, 61],
 ['C', 60, 4, 81],
 ['C', 55, 24, 82],
 ['C', 58, 25, 98],
 ['C', 20, 0, 121],
 ['I', 64, 19, 136],
 ['C', 2, 26, 143],
 ['I', 8, 19, 165],
 ['I', 51, 3, 169],
 ['C', 61, 1, 173],
 ['C', 58, 18, 176],
 ['C', 59, 20, 179],
 ['C', 36, 2, 186],
 ['D', 29, 198],
 ['I', 64, 19, 202],
 ['C', 8, 8, 214],
 ['C', 61, 12, 214],
 ['C', 60, 20, 221],
 ['I', 43, 16, 223],
 ['C', 0, 10, 226],
 ['C', 60, 15, 232],
 ['C', 62, 5, 235],
 ['I', 63, 18, 259],
 ['C', 3, 5, 263],
 ['I', 20, 6, 275],
 ['C', 46, 17, 308],
 ['C', 64, 16, 309],
 ['C', 3, 14, 311],
 ['C', 62, 19, 320],
 ['I', 29, 4, 323],
 ['D', 24, 325],
 ['C', 56, 5, 335],
 ['I', 26, 17, 335],
 ['C', 17, 13, 340],
 ['I', 47, 15, 340]]

XOR -> EQU:

340 Generations

Mutations:
Copy: 27
Insert: 11
Delete: 2
TOTAL: 40

EQU of length 79

Does contain the "move" instruction

[10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]

In [13]:
"""
Recreating the program at all generations

After running this script, lineage[i] contains the program at generation i,
generation 0 being the original ancestor, generation 1 its child and so on
"""

generations = mutations[-1][-1] + 1

start = em.ancestor.copy()
mutations = em.mutations.copy()

temp = start.copy()

lineage = [start]

for i in range(generations):
    
    for mutation in mutations:
        
        if mutation[-1] == i:
            
            if mutation[0] == "C":
                temp[mutation[1]] = mutation[2]
        
            if mutation[0] == "I":
                temp.insert(mutation[1], mutation[2])
        
            if mutation[0] == "D":
                del temp[mutation[1]]
                
    lineage.append(temp.copy())

In [14]:
for specimen in lineage:
    print(specimen)
    print("\n")

[16, 20, 2, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 19, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0, 6, 0, 6, 1, 6, 2, 20, 19, 25, 2, 0, 17, 21, 0, 1]


[16, 20, 23, 0, 21, 2, 18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0

In [4]:
world = World(2, notify_ = False, cm_prob = 0, ins_prob = 0, del_prob = 0)
world.place_custom([10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1])

In [6]:
world.rates

array([[   79, 20224],
       [  316,   316]])

In [67]:
experiment = Experiment(start_organism="not", target_function="nor", ins_prob = 0,\
                        del_prob = 0, cm_prob = 0,N=3, stat_cycles = 100, instruction_set = "custom")

In [68]:
experiment.world.pool.get()

array([[<DigitalOrganism.CPUEmulator object at 0x000002AE875879A0>,
      dtype=object)

In [69]:
experiment.world.rates

array([[50, 50, 50],
       [50, 50, 50],
       [50, 50, 50]])

In [226]:
killer = [16, 20, 2, 0, 21] + [2] + [18, 0, 18, 2, 7, 0, 7, 2, 8, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15,2, 6, 1, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 1, 6, 0, 15, 2, 8, 6, 1, 15, 0, 18, 0]\
                            + [6, 0, 6, 1, 6, 2] + [20, 19, 25, 2, 0, 17, 21, 0, 1]

killer = [10, 20, 26, 14, 21, 2, 18, 0, 8, 18, 2, 7, 0, 7, 2, 8, 14, 13, 12, 2, 6, 7, 0, 6, 6, 0, 17, 15, 2, 4, 6, 7, 2, 8, 6, 0, 6, 2, 14, 2, 12, 2, 7, 0, 6, 16, 1, 15, 17, 0, 15, 2, 8, 6, 3, 3, 15, 0, 5, 0, 24, 18, 20, 15, 19, 5, 16, 20, 19, 19, 19, 19, 25, 2, 0, 17, 21, 0, 1]

In [234]:
world = World(10, notify_ = False, cm_prob = 0, ins_prob = 0, del_prob = 0)
world.place_custom(killer)

In [237]:
world.schedule(1000)
world.rates

array([[   79,    79,   316,    79,  5056,   316, 20224,   316,    79,
           79],
       [   79,    79,   316,    79, 20224,    79,  5056,    79,  5056,
          316],
       [   79,   316,    79,   316,   316, 20224,    79,    79,  5056,
           79],
       [  316,    79,    79, 20224,   316,    79,    79,    79,    79,
        20224],
       [   79,   316,  5056,    79,   316, 20224,  5056,    79,   316,
          316],
       [   79,    79,    79,   316, 20224,   316,   316,    79, 20224,
           79],
       [  316,    79,   316,    79,   316,    79,    79,  5056,   316,
           79],
       [20224, 20224, 20224,   316,    79,    79,    79,    79,   316,
           79],
       [  316,   316,   316, 20224,   316,    79, 20224, 20224,    79,
           79],
       [  316,    79,  5056,   316,   316, 20224,  5056,  5056, 20224,
        20224]])